src/sentry/middleware/ratelimit.py

```diff
@@ -15,80 +16,122 @@
     "{limit} requests in {window} seconds"
 )
 
+rate_limit_logger = logging.getLogger("sentry.django.ratelimiter")
 
-class RatelimitMiddleware(MiddlewareMixin):
-    """Middleware that applies a rate limit to every endpoint."""
 
-    def process_view(self, request: Request, view_func, view_args, view_kwargs) -> Response | None:
-        """Check if the endpoint call will violate."""
+class RatelimitMiddleware:
+    """Middleware that applies a rate limit to every endpoint.
+    See: https://docs.djangoproject.com/en/4.0/topics/http/middleware/#writing-your-own-middleware
+    """
+
+    def __init__(self, get_response: Callable[[Request], Response]):
+        self.get_response = get_response
+
+    def __call__(self, request: Request) -> Response:
+        response, rate_limit_data = self.process_request(request)
+        # Hit the endpoint
+        if hasattr(rate_limit_data, "rate_limit_metadata"):
+            request.rate_limit_metadata = rate_limit_data.rate_limit_metadata
+        if not response:
+            response = self.get_response(request)
+
+        # Process the response
+        self.add_headers(response, request)
+        if hasattr(rate_limit_data, "rate_limit_key") and hasattr(
+            rate_limit_data, "rate_limit_uid"
+        ):
+            finish_request(rate_limit_data.rate_limit_key, rate_limit_data.rate_limit_uid)
+        return response
+
+    def _debug_log(self, request):
+        metrics = {}
+        metrics["path"] = request.path
+        metrics["resolved"] = str(resolve(request.path))
+        rate_limit_logger.info("sentry.ratelimiter", extra=metrics)
+
+    def process_request(self, request: Request) -> Tuple[HttpResponse | None, RateLimitData]:
+        rate_limit_data: RateLimitData = None
+        rate_limit_uid = None
+        rate_limit_key = None
+        rate_limit_metadata = None
+        response = None
+
+        # First, check if the endpoint call will violate.
         try:
-            # TODO: put these fields into their own object
-            request.will_be_rate_limited = False
-            request.rate_limit_category = None
-            request.rate_limit_uid = uuid.uuid4().hex
-            request.rate_limit_key = get_rate_limit_key(view_func, request)
-            if request.rate_limit_key is None:
-                return
-            category_str = request.rate_limit_key.split(":", 1)[0]
-            request.rate_limit_category = category_str
+            rate_limit_uid = uuid.uuid4().hex
+            # in deprecated django middleware the view function is passed in to process_view
+            # but we can still access it this way through the request
+            self._debug_log(request)
+            view_func = resolve(request.path).func
+            rate_limit_key = get_rate_limit_key(view_func, request)
+            if rate_limit_key is None:
+                return response, rate_limit_data
 
+            category_str = rate_limit_key.split(":", 1)[0]
```

Comment:
Since this is always going to be read as a `RateLimitCategory`, I'd do:
```py
category_str = RateLimitCategory(rate_limit_key.split(":", 1)[0])
```